### PR TITLE
ci: use asdf to manage tool versions

### PR DESCRIPTION
**Pull request by ggilmore**
Fixes: https://github.com/sourcegraph/sourcegraph/issues/6862

Depends on: https://github.com/sourcegraph/infrastructure/pull/1740

This only handles Go, Yarn, and Node for now, but we can install other things like hadolint, postgres, comby, etc. in the future with asdf.

---

The only slightly annoying thing is that you have to run `asdf reshim golang` any time you `go get` something: https://github.com/kennyp/asdf-golang#when-using-go-get

This is because asdf has no way of knowing that your `go` command ended up doing something that needs to have a new `shim` created for it.

### CODE REVIEW

File: .tool-versions [hunk 1]
```diff
@@ -0,0 +1,2 @@
+golang 1.13.4
+nodejs 12.10.0
```
Comment (felixfbecker) [inline .tool-versions]: I read in the readme that asdf supports nvm's `.nvmrc`, which more people probably use than asdf. Renovate is also only able to renovate `node` versions in package.json `engines` and `.nvmrc`. I'd therefor propose we use `.nvmrc` for the Node version

Comment (felixfbecker) [reply .tool-versions]: Also, this should be 13 now :)

Comment (ggilmore) [reply .tool-versions]: I just used the current node version that's in the buildkite agent

Comment (ggilmore) [reply .tool-versions]: What is the exact version number that I should be using?

Comment (unknwon) [reply .tool-versions]: 13.2.0 is the latest I think.

File: enterprise/dev/ci/ci/pipeline.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ func GeneratePipeline(c Config) (*bk.Pipeline, error) {
 		bk.Env("ENTERPRISE", "1"),
 		bk.Env("COMMIT_SHA", c.commit),
 		bk.Env("DATE", c.now.Format(time.RFC3339)),
+		bk.Cmd("asdf install"),
```
Comment (keegancsmith) [inline enterprise/dev/ci/ci/pipeline.go]: not all steps need go, and not all steps need node. But most do, so this seems good for now.

File: .tool-versions (hunk 1 above)
Comment (ggilmore) [reply .tool-versions]: @felixfbecker  It seems like there are some issues with node `13.2.0`, so I'm sticking with `12.10.0` for now: https://buildkite.com/sourcegraph/sourcegraph/builds/49994#

Comment (felixfbecker) [reply .tool-versions]: Hmm that looks like a flake unrelated to Node 13... I don't get that locally. Tried rebuilding that job but the build got cancelled by the revert

Comment (ggilmore) [reply .tool-versions]: Feel free to rebuild that older commit, but I've seen it a few times now in buildkite - flipping between 12 and 13 seems to cause this

Comment (ggilmore) [reply .tool-versions]: Note that the buildkite agents are still using node 12. Perhaps there is some other state on your machine that makes it work with node 13 that needs to be added to buildkite?

Comment (ggilmore) [reply .tool-versions]: @felixfbecker I'm going to move forward with still using node 12. It feels like upgrading the node version should be handled in a separate PR anyway.

Comment (felixfbecker) [reply .tool-versions]: Here's a case of the same error with Node 12: https://buildkite.com/sourcegraph/sourcegraph/builds/50014#7220ce3b-6ccd-4dc8-b572-a88ef1f77248/84-1913

Comment (felixfbecker) [reply .tool-versions]: (okay with keeping Node 12 though)

**Comment by unknwon** [comment]
> The only slightly annoying thing is that you have to run `asdf reshim golang` any time you `go get` something: [kennyp/asdf-golang#when-using-go-get](https://github.com/kennyp/asdf-golang#when-using-go-get)

Does this only affects CI and/or in dev environment?

**Comment by ggilmore** [comment]
@unknwon This would affect both (locally if you you use asdf - although it’s probably straightforward to write a helper function to script this)

**Comment by keegancsmith** [comment]
> The only slightly annoying thing is that you have to run asdf reshim golang any time you go get something: https://github.com/kennyp/asdf-golang#when-using-go-get
> This is because asdf has no way of knowing that your go command ended up doing something that needs to have a new shim created for it.

Is GOBIN still on the PATH? If so then this shouldn't be an issue?

**Comment by keegancsmith** [comment]
This seems like a useful plugin for `direnv` users https://github.com/asdf-community/asdf-direnv

As an alternative to asdf I have heard of a few dev shops using nix. But that is likely too heavy weight for what we want right now, while asdf seems nice and lightweight.

**Comment by ggilmore** [comment]
> > The only slightly annoying thing is that you have to run asdf reshim golang any time you go get something: [kennyp/asdf-golang#when-using-go-get](https://github.com/kennyp/asdf-golang#when-using-go-get)
> > This is because asdf has no way of knowing that your go command ended up doing something that needs to have a new shim created for it.
> 
> Is GOBIN still on the PATH? If so then this shouldn't be an issue?

Each binary now needs it's own shim to know what version of the binary to use (i.e. the go version that it was built with).

**Comment by ggilmore** [comment]
> I was surprised the pre-command took 5s to run in this PR (on the one step I looked at). It looks like it is downloading 80mb from somewhere, but it doesn't say what. I assume it is go since it didn't give a message about go. I know 5s isn't that much, but ideally we would want this step to be a noop in the general case.
> 
> ```
> ~~~ Running local pre-command hook
> [2019-12-04T22:27:41Z] �[90m$�[0m .buildkite/hooks/pre-command
> [2019-12-04T22:27:41Z] Installing asdf dependencies as defined in '/buildkite/builds/buildkite-agent-9957fc659-hsxb2-1/sourcegraph/sourcegraph/.tool-versions':
> [2019-12-04T22:27:41Z] Platform 'linux' supported!
> [2019-12-04T22:27:41Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
> [2019-12-04T22:27:41Z]                                  Dload  Upload   Total   Spent    Left  Speed
> [2019-12-04T22:27:43Z] 
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>  17  114M   17 19.9M    0     0  47.7M      0  0:00:02 --:--:--  0:00:02 47.6M
> 100  114M  100  114M    0     0  81.8M      0  0:00:01  0:00:01 --:--:-- 81.8M
> [2019-12-04T22:27:47Z] nodejs 12.10.0 is already installed
> [2019-12-04T22:27:47Z] yarn 1.17.3 is already installed
> [2019-12-04T22:27:47Z] �[90m# OLDPWD changed�[0m
> ```

@keegancsmith 

How would we speed this up? We'd need to bake the version of all the tools that we're using in the buildkite agent, which is the same situation that we're in now.

**Comment by keegancsmith** [comment]
> How would we speed this up? We'd need to bake the version of all the tools that we're using in the buildkite agent, which is the same situation that we're in now.

Yes bake them in for the common case. The main point of using this tool is it is possible for us to build with older versions of go when cutting a patch release. However, it would be ideal to not pay a performance penalty in the common case.

If `asdf` had a way to store cached files, that would be ideal! IE in that case I'd expect it to only take 1s since it could avoid the network.

**Pull request merged**